fearless/ApplicationLayer/Services/WalletBalanceSubscriptionAdapter.swift

```diff
@@ -0,0 +1,470 @@
+
+protocol WalletBalanceFetcherHandler: AnyObject {
+    func handle(result: WalletBalancesResult)
+}
+
+protocol WalletBalanceSubscriptionAdapterProtocol {
+    /// Collects and counts all the information for `WalletBalance`
+    /// - Parameters:
+    ///   - walletId: Balance for specific wallet Id. If nil calculating for all meta accounts
+    ///   - completion: Called when WalletBalance will calculated
+    ///   - queue: The queue to which the result will be delivered
+    func subsctibeWalletBalance(
+        walletId: String?,
+        deliverOn queue: DispatchQueue?,
+        handler: WalletBalanceFetcherHandler
+    )
+}
+
+enum WalletBalanceError: Error {
+    case accountMissing
+    case chainsMissing
+    case `internal`
+}
+
+final class WalletBalanceSubscriptionAdapter: WalletBalanceSubscriptionAdapterProtocol, PriceLocalStorageSubscriber {
+    // MARK: - PriceLocalStorageSubscriber
+
+    let priceLocalSubscriptionFactory: PriceProviderFactoryProtocol
+
+    // MARK: - Private properties
+
+    private var pricesProvider: AnySingleValueProvider<[PriceData]>?
+    private lazy var walletBalanceBuilder = {
+        WalletBalanceBuilder()
+    }()
+
+    private let metaAccountRepository: AnyDataProviderRepository<MetaAccountModel>
+    private let chainRepository: AnyDataProviderRepository<ChainModel>
+    private let operationQueue: OperationQueue
+    private let eventCenter: EventCenterProtocol
+    private let logger: Logger
+
+    private var deliverQueue: DispatchQueue?
+    private weak var delegate: WalletBalanceFetcherHandler?
+
+    private lazy var accountInfosAdapters: [String: AccountInfoSubscriptionAdapter] = [:]
+    private lazy var accountInfos: [ChainAssetKey: AccountInfo?] = [:]
+    private lazy var chainAssets: [ChainAsset] = []
+    private lazy var metaAccounts: [MetaAccountModel] = []
+    private lazy var prices: [PriceData] = []
+
+    // MARK: - Constructor
+
+    init(
+        metaAccountRepository: AnyDataProviderRepository<MetaAccountModel>,
+        priceLocalSubscriptionFactory: PriceProviderFactoryProtocol,
+        chainRepository: AnyDataProviderRepository<ChainModel>,
+        operationQueue: OperationQueue,
+        eventCenter: EventCenterProtocol,
+        logger: Logger
+    ) {
+        self.priceLocalSubscriptionFactory = priceLocalSubscriptionFactory
+        self.metaAccountRepository = metaAccountRepository
+        self.chainRepository = chainRepository
+        self.operationQueue = operationQueue
+        self.eventCenter = eventCenter
+        self.logger = logger
+        eventCenter.add(observer: self, dispatchIn: .global())
+    }
+
+    // MARK: - Public methods
+
+    func subsctibeWalletBalance(
+        walletId: String?,
+        deliverOn queue: DispatchQueue?,
+        handler: WalletBalanceFetcherHandler
+    ) {
+        deliverQueue = queue
+        delegate = handler
+
+        if let identifier = walletId {
```

Comment:
please divide on two methods. walletId is some case of filtering. imagine you have 4 filter cases - then you'll have 4 else if conditions here